cfme/markers/upgrade.py

```diff
@@ -0,0 +1,63 @@
+"""Marker definitions for upgrade testing"""
+import pytest
+
+
+# Add Marker
+def pytest_configure(config):
+    config.addinivalue_line("markers", "post_upgrade: Mark test for post upgrade testing")
+
+
+# Filtering options
+def pytest_addoption(parser):
+    group = parser.getgroup("cfme")
+    group.addoption(
+        "--upgrade-appliance",
+        dest="upgrade_appliance",
+        action="store_true",
+        default=False,
+        help="Upgrade an appliance before tests are run.",
+    )
+    group.addoption(
+        "--upgrade-to",
+        dest="upgrade_to",
+        action="store",
+        default="5.11.z",
+        help="Supported versions 5.9.z, 5.10.z, 5.11.z (.z means latest and default is 5.11.z)",
+    )
+
+
+# Test collection modifier
+def pytest_collection_modifyitems(session, config, items):
+    """Provide filtering of test case collection based on the CLI options"""
+    post_upgrade = config.getoption("upgrade_appliance")
```

Comment:
I'm not sure this is the behavior we actually want.

There are two separate things being combined and made interdependent here:
1. post-upgrade pytest option, which should trigger the actual upgrade of appliances
2. the marker filtering of test collection

The first is obviously not provided by this function, we're just tying whether or not to modify collection based on it.  I don't think it should be linked. We should be able to run _all_ tests along with filter the collection of them.

You can just use `pytest -m post_upgrade` to filter test collection on the marker - without any of this code, its built into pytest `-m` option.  The only downside of using this is the uncollection metrics that get printed just include non-matched markers in the 'uncollectif' grouping.

```
pytest --collect-only -m junk cfme/tests/webui/test_general_ui.py

==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.7.4, pytest-3.4.1, py-1.8.0, pluggy-0.6.0
rootdir: /home/mshriver/repos/integration_tests, inifile:
plugins: polarion-collect-0.21.0, manageiq-integration-tests-17.61.1.dev9+g5b40aa1e1
collected 22 items                                                                                                                                                                           
Uncollection Stats:
 manual: 10
 uncollectif: 5
 0 tests left after all uncollections

```

And with a slight modification, adding a '@pytest.mark.junk' marker to a test function, with no other code changes:

```
pytest --collect-only -m junk cfme/tests/webui/test_general_ui.py

==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.7.4, pytest-3.4.1, py-1.8.0, pluggy-0.6.0
rootdir: /home/mshriver/repos/integration_tests, inifile:
plugins: polarion-collect-0.21.0, manageiq-integration-tests-17.61.1.dev9+g5b40aa1e1
collected 22 items                                                                                                                                                                           
<Module 'cfme/tests/webui/test_general_ui.py'>
  <Function 'test_add_provider_trailing_whitespaces[virtualcenter]'>
Uncollection Stats:
 manual: 10
 uncollectif: 5
 1 tests left after all uncollections

```

So the only value-add in defining the hookimpl that controls collection is that we can specify a particular uncollection_stat field for it, which I'm not sure we need.  The check against upgrade_appliance should not also impact whether or not the marked tests are included.  We should be able to: 1) collect post_upgrade tests without including the upgrade option and 2) include the upgrade option without limiting test collection

Reply:
@mshriver the behavior you want actually I removed https://github.com/ManageIQ/integration_tests/pull/9510/commits/164cc21f77f50f6b69e3ef1bd7c3b7f6c6c03cf2.

removing `modifier` and lets collect test with `-m`